Mathlib/LinearAlgebra/RootSystem/Finite/CanonicalBilinear.lean

```diff
@@ -0,0 +1,139 @@
+
+/-!
+# The canonical bilinear form on a finite root pairing
+Given a finite root pairing, we define a canonical map from weight space to coweight space, and the
+corresponding bilinear form. This form is symmetric and Weyl-invariant, and if the base ring is
+linearly ordered, then the form is root-positive, positive-semidefinite on the weight space, and
+positive-definite on the span of roots.
+From these facts, it is easy to show that Coxeter weights in a finite root pairing are bounded
+above by 4. Thus, the pairings of roots and coroots in a crystallographic root pairing are
+restricted to a small finite set of possibilities.
+Another application is to the faithfulness of the Weyl group action on roots, and finiteness of the
+Weyl group.
+
+## Main definitions:
+ * `Polarization`: A distinguished linear map from the weight space to the coweight space.
+ * `CanonicalBilinear` : The bilinear form on weight space corresponding to `Polarization`.
+
+## References:
+ * SGAIII Exp. XXI
+ * Bourbaki, Lie groups and Lie algebras
+
+## Main results:
+ * `polarization_self_sum_of_squares` : The inner product of any weight vector is a sum of squares.
+ * `canonicalBilinear_reflection_reflection_apply` : `CanonicalBilinear` is invariant with respect
+   to reflections.
+ * `canonicalBilinear_self_smul_coroot`: The inner product of a root with itself times the
+   corresponding coroot is equal to two times Polarization applied to the root.
+
+## TODO (possibly in other files)
+ * Positivity and nondegeneracy
+ * Weyl-invariance
+ * Faithfulness of Weyl group action, and finiteness of Weyl group, for finite root systems.
+ * Relation to Coxeter weight.  In particular, positivity constraints for finite root pairings mean
+  we restrict to weights between 0 and 4.
+-/
+
+open Set Function
+open Module hiding reflection
+
+noncomputable section
+
+variable {ι R M N : Type*}
+
+lemma isSumSq_of_sum_of_squares [Mul R] [AddCommMonoid R] (s : Finset ι) (f : ι → R) :
+    IsSumSq (∑ i ∈ s, f i * f i) := by
+  induction s using Finset.cons_induction with
+  | empty =>
+    simpa only [Finset.sum_empty] using IsSumSq.zero
+  | cons i s his h =>
+    simp only [Finset.sum_cons]
+    exact IsSumSq.sq_add (f i) (∑ i ∈ s, f i * f i) h
+--#find_home! isSumSq_of_sum_of_squares --returns this file
+
+namespace RootPairing
+
+section CommRing
+
+variable [Fintype ι] [CommRing R] [AddCommGroup M] [Module R M] [AddCommGroup N] [Module R N]
+(P : RootPairing ι R M N)
+
+/-- An invariant linear map from weight space to coweight space. -/
+@[simps]
+def Polarization : M →ₗ[R] N where
+  toFun m := ∑ (i : ι), P.toPerfectPairing m (P.coroot i) • (P.coroot i)
+  map_add' x y := by
+    simp only [← toLin_toPerfectPairing, map_add, PerfectPairing.toLin_apply, LinearMap.add_apply,
+      add_smul, Finset.sum_add_distrib]
+  map_smul' r x := by
+    simp only [← toLin_toPerfectPairing, map_smul, LinearMap.smul_apply, RingHom.id_apply,
+      Finset.smul_sum, smul_assoc]
```

Comment:
It's rare that we have to prove linearity manually in situations like this.

I've thought for some time that in `RootSystem.Defs` we should introduce:
```lean
abbrev root' (i : ι) : Dual R N := P.toPerfectPairing (P.root i)
abbrev coroot' (i : ι) : Dual R M := P.toPerfectPairing.flip (P.coroot i)
```
(they're not necessary, just convenient, especially here) and then define this as:
```suggestion
def Polarization : M →ₗ[R] N :=
  ∑ i, LinearMap.toSpanSingleton R N (P.coroot i) ∘ₗ P.coroot' i
 
 @[simp] lemma Polarization_apply (x : M) :
    P.Polarization x = ∑ i, P.toPerfectPairing x (P.coroot i) • P.coroot i := by
  simp [Polarization]
```

Reply:
I shouldn't really write this comment because I haven't had time to think about it and I only have a moment before I start work.

Anyway it occurred to me this morning that in theory one could also define:
```lean
def CoPolarization : N →ₗ[R] M :=
  ∑ i, LinearMap.toSpanSingleton R M (P.root i) ∘ₗ P.root' i
```
and that presumably `Polarization` and `CoPolarization` are mutually inverse and this is equivalent to non-degeneracy of the canonical bilinear forms.

If that's right we might want to delay defining `Polarization` until after non-degeneracy has been defined so that we can instead just define `Polarization` but as a `LinearEquiv` (and no `CoPolarization`).

Reply:
Yes to your edit. We basically never get an equivalence when 2 is not invertible.  For sl2, \alpha is sent to 4\alpha^*.

Reply:
I should mention that I only get nondegeneracy when I have `[LinearOrderedCommRing R]`.